### PR TITLE
Pyupgrade part 2

### PR DESCRIPTION
**Pull request by chris-rands**
See PR #2450 for background. Here I ran `pyupgrade --keep-percent-format --py3-plus` on the relevant `*py` files. This does not introduce f-strings since that would require the `--py36-plus` flag.

Note, the changes include python 3 changes (e.g. `yield from`,  or `IOError` merged into `OSError ` etc.), but also other changes not related to the python version specifically like removing redundant parentheses and string formatting syntax. See [their github ](https://github.com/asottile/pyupgrade) for a full list of what is checked. Hopefully this is not too large to review

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

**Comment by peterjc** [comment]
Still a lot of changes to eyeball - the largest single group looks like ``OSError`` to me, again something we could fix alone with sed (or your favourite global search-and-replace tool).

There are a couple of cases of dropping the ``u`` prefix on unicode literals - I must have missed a few on #2442.

**Comment by peterjc** [comment]
Flake8 is interesting...

```
$ flake8
./Bio/SearchIO/_model/query.py:233:24: B301 Python 3 does not include `.iter*` methods on dictionaries. Remove the `iter` prefix from the method name. For Python 2 compatibility, prefer the Python 3 equivalent unless you expect the size of the container to be large or unbounded. Then use `six.iter*` or `future.utils.iter*`.
./Bio/SearchIO/_model/query.py:241:24: B301 Python 3 does not include `.iter*` methods on dictionaries. Remove the `iter` prefix from the method name. For Python 2 compatibility, prefer the Python 3 equivalent unless you expect the size of the container to be large or unbounded. Then use `six.iter*` or `future.utils.iter*`.
./Bio/Align/substitution_matrices/__init__.py:215:58: E127 continuation line over-indented for visual indent
./Bio/codonalign/__init__.py:561:50: E127 continuation line over-indented for visual indent
```

**Comment by peterjc** [comment]
OK, there were ``# noqa: B301`` lines in ``Bio/SearchIO/_model/query.py`` which got lost - over to @bow to assess what needs updating there now we've dropped Python 2 support.

**Comment by peterjc** [comment]
The string formatting changes from ``"e.g. {0}" % x`` to ``"e.g. {}" % x`` do seem a lot of churn for minimal benefit... especially if they would then get replaced with f-strings later on. I wasn't expecting this when we discussed splitting up #2450  into chunks to aid review.

File: Bio/AlignIO/__init__.py [hunk 1]
```diff
@@ -388,8 +388,7 @@ def parse(handle, format, seq_count=None, alphabet=None):
 
         # TODO: As of Python 3.3, can write "yield from i" instead. See PEP380.
```
Comment (mdehoon) [inline Bio/AlignIO/__init__.py]: comment can be removed

File: Bio/AlignIO/__init__.py [hunk 2]
```diff
@@ -388,8 +388,7 @@ def parse(handle, format, seq_count=None, alphabet=None):
...
         # For loop imposes some overhead... wait until we drop Python 2.7 to fix it.
```
Comment (mdehoon) [inline Bio/AlignIO/__init__.py]: comment can be removed

File: Scripts/Restriction/ranacompiler.py [hunk 3]
```diff
@@ -463,9 +463,9 @@ def build_dict(self):
             else:
                 enzlst = []
                 tydct = dict(typestuff.__dict__)
-                tydct = dict(  # noqa: C404
-                    [(k, v) for k, v in tydct.items() if k in commonattr]
-                )
+                tydct = {  # noqa: C404
+                    k: v for k, v in tydct.items() if k in commonattr
```
Comment (mdehoon) [inline Scripts/Restriction/ranacompiler.py]: this can now be on one line

**Comment by chris-rands** [comment]
Thanks both. @peterjc I agree with your comment about some string formatting being minimal benefit. When I have more time I can think about this. In the mean time, you can merge your own PRs and I'll rebase this and/or #2450 later

**Comment by peterjc** [comment]
Rebased to current master...

File: Bio/Entrez/Parser.py [hunk 4]
```diff
@@ -328,7 +328,7 @@ def read(self, handle):
         if hasattr(handle, "closed") and handle.closed:
             # Should avoid a possible Segmentation Fault, see:
             # http://bugs.python.org/issue4877
-            raise IOError("Can't parse a closed handle")
+            raise OSError("Can't parse a closed handle")
```
Comment (bow) [inline Bio/Entrez/Parser.py]: I'm not sure if this should be changed here, for the record, but would it be an idea to change this to `ValueError` instead?

Thea reason is `read()` on a closed file handle raises a `ValueError` on Python 3 (actually on Python 2 as well):

```
Python 3.8.1 (default, Dec 21 2019, 20:57:38) 
[GCC 9.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> f = open(<fname>, "r")
>>> f.close()
>>> f.read()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: I/O operation on closed file.
```
```

File: Bio/Phylo/PAML/baseml.py [hunk 5]
```diff
@@ -33,7 +33,7 @@ def __init__(self, alignment=None, tree=None, working_dir=None, out_file=None):
         Paml.__init__(self, alignment, working_dir, out_file)
         if tree is not None:
             if not os.path.exists(tree):
-                raise IOError("The specified tree file does not exist.")
+                raise OSError("The specified tree file does not exist.")
```
Comment (bow) [inline Bio/Phylo/PAML/baseml.py]: It seems that `FileNotFoundError` is a better exception here as it's an OSError subclass that precisely describes the error, e.g.:

```
Python 3.8.1 (default, Dec 21 2019, 20:57:38) 
[GCC 9.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> f = open("nonexistent", "r")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
FileNotFoundError: [Errno 2] No such file or directory: 'nonexistent'
```

File: Bio/Phylo/PAML/baseml.py [hunk 6]
```diff
@@ -104,7 +104,7 @@ def read_ctl_file(self, ctl_file):
         """Parse a control file and load the options into the Baseml instance."""
         temp_options = {}
         if not os.path.isfile(ctl_file):
-            raise IOError("File not found: %r" % ctl_file)
+            raise OSError("File not found: %r" % ctl_file)
```
Comment (bow) [inline Bio/Phylo/PAML/baseml.py]: Same as above, this should be `FileNotFoundError`

File: Bio/Phylo/PAML/baseml.py [hunk 7]
```diff
@@ -177,7 +177,7 @@ def run(self, ctl_file=None, verbose=False, command="baseml", parse=True):
         if self.tree is None:
             raise ValueError("Tree file not specified.")
         if not os.path.exists(self.tree):
-            raise IOError("The specified tree file does not exist.")
+            raise OSError("The specified tree file does not exist.")
```
Comment (bow) [inline Bio/Phylo/PAML/baseml.py]: This should be another `FileNotFoundError` instance. I'll continue adding these comments on the same errors below to as sign-posts when updating.

File: Bio/Phylo/PAML/baseml.py [hunk 8]
```diff
@@ -188,7 +188,7 @@ def read(results_file):
     """Parse a BASEML results file."""
     results = {}
     if not os.path.exists(results_file):
-        raise IOError("Results file does not exist.")
+        raise OSError("Results file does not exist.")
```
Comment (bow) [inline Bio/Phylo/PAML/baseml.py]: This should be a `FileNotFoundError` instance.

File: Bio/Phylo/PAML/codeml.py [hunk 9]
```diff
@@ -35,7 +35,7 @@ def __init__(self, alignment=None, tree=None, working_dir=None, out_file=None):
         Paml.__init__(self, alignment, working_dir, out_file)
         if tree is not None:
             if not os.path.exists(tree):
-                raise IOError("The specified tree file does not exist.")
+                raise OSError("The specified tree file does not exist.")
```
Comment (bow) [inline Bio/Phylo/PAML/codeml.py]: This should be a `FileNotFoundError` instance.

File: Bio/Phylo/PAML/codeml.py [hunk 10]
```diff
@@ -102,7 +102,7 @@ def read_ctl_file(self, ctl_file):
         """Parse a control file and load the options into the Codeml instance."""
         temp_options = {}
         if not os.path.isfile(ctl_file):
-            raise IOError("File not found: %r" % ctl_file)
+            raise OSError("File not found: %r" % ctl_file)
```
Comment (bow) [inline Bio/Phylo/PAML/codeml.py]: This should be a `FileNotFoundError` instance.

File: Bio/Phylo/PAML/codeml.py [hunk 11]
```diff
@@ -187,7 +187,7 @@ def run(self, ctl_file=None, verbose=False, command="codeml", parse=True):
         if self.tree is None:
             raise ValueError("Tree file not specified.")
         if not os.path.exists(self.tree):
-            raise IOError("The specified tree file does not exist.")
+            raise OSError("The specified tree file does not exist.")
```
Comment (bow) [inline Bio/Phylo/PAML/codeml.py]: This should be a `FileNotFoundError` instance.

File: Bio/Phylo/PAML/codeml.py [hunk 12]
```diff
@@ -198,7 +198,7 @@ def read(results_file):
     """Parse a CODEML results file."""
     results = {}
     if not os.path.exists(results_file):
-        raise IOError("Results file does not exist.")
+        raise OSError("Results file does not exist.")
```
Comment (bow) [inline Bio/Phylo/PAML/codeml.py]: This should be a `FileNotFoundError` instance.

File: Bio/Phylo/PAML/yn00.py [hunk 13]
```diff
@@ -67,7 +67,7 @@ def read_ctl_file(self, ctl_file):
         """Parse a control file and load the options into the yn00 instance."""
         temp_options = {}
         if not os.path.isfile(ctl_file):
-            raise IOError("File not found: %r" % ctl_file)
+            raise OSError("File not found: %r" % ctl_file)
```
Comment (bow) [inline Bio/Phylo/PAML/yn00.py]: This should be a `FileNotFoundError` instance.

File: Bio/Phylo/PAML/yn00.py [hunk 14]
```diff
@@ -121,7 +121,7 @@ def read(results_file):
     """Parse a yn00 results file."""
     results = {}
     if not os.path.exists(results_file):
-        raise IOError("Results file does not exist.")
+        raise OSError("Results file does not exist.")
```
Comment (bow) [inline Bio/Phylo/PAML/yn00.py]: This should be a `FileNotFoundError` instance.

**Comment by peterjc** [comment]
Using ``FileNotFoundError`` makes sense - that was new in Python 3

File: Bio/Entrez/Parser.py (hunk 4 above)
Comment (mdehoon) [reply Bio/Entrez/Parser.py]: We don't need to check for closed files at all. See the original issue in http://bugs.python.org/issue4877 ; this was fixed in 2011.

Comment (peterjc) [reply Bio/Entrez/Parser.py]: So that if-clause can be removed entirely?

**Comment by peterjc** [comment]
Close in favour of #2480?

**Comment by peterjc** [comment]
With #2480 merged, I think we you were to repeat this now, it would just be f-string changes, and not too onerous to review?

**Comment by peterjc** [comment]
Sorry, just see your other comment about f-strings: https://github.com/biopython/biopython/pull/2450#issuecomment-570091366

**Pull request closed without merging**